### PR TITLE
added a boolean to modify the current previous song handler

### LINKED ISSUE #347: Way to extend the playPrev and playNext functions
### Description
Currently, we cannot add our logic to the default methods of playing the prev track or next track. 

https://github.com/lijinke666/react-music-player/blob/master/src/index.js#L1539-L1551

Let's say, I click on the previous button. I want to check that if the `AudioInstance` 's currentTime is greater than `0`, then restart the current track and if not, play the previous track. Is there any way to configure this?

### Expect Api Name
`onAudioPrev`/ `onAudioNext`

### PR DESCRIPTION
**Pull request by aniketbiswas21**
Fixes #347 

I exposed a boolean `restartCurrentOnPrev` that if set to `true`, restarts the current song if the current time of the song is greater than 1 second else plays the previous song on the list(the way it was handled earlier).

`restartCurrentonPrev` defaults to `false` => which retains the current behaviour.

### CODE REVIEW

**Comment by aniketbiswas21** [comment]
@lijinke666 Please take a look!

File: index.d.ts [hunk 1]
```diff
@@ -234,6 +234,7 @@ export interface ReactJkMusicPlayerProps {
     fadeIn?: number
     fadeOut?: number
   }
+  restartCurrentOnPrev: Boolean
```
Comment (lijinke666) [inline index.d.ts:237]: restartCurrentOnPrev?: boolean

File: src/index.js [hunk 2]
```diff
@@ -1460,7 +1462,7 @@ export default class ReactJkMusicPlayer extends PureComponent {
 
   // isNext true 下一首  false
   handlePlay = (playMode, isNext = true) => {
-    const { playId, audioLists } = this.state
+    const { playId, audioLists, restartCurrentOnPrev } = this.state
```
Comment (lijinke666) [inline src/index.js:1465]: why not use from props ?

File: src/index.js [hunk 3]
```diff
@@ -1490,11 +1504,30 @@ export default class ReactJkMusicPlayer extends PureComponent {
...
           this.audioListsPlay(audioLists[currentPlayIndex + 1][PLAYER_KEY])
         } else {
+          if (restartCurrentOnPrev && currentPlayIndex === 0) {
+            if (this.audio.currentTime <= 1) {
+              return this.audioListsPlay(
+                audioLists[audioListsLen - 1][PLAYER_KEY],
+              )
+            }
+            this.audio.currentTime = 0
+            return
```
Comment (lijinke666) [inline src/index.js:1514]: Please don't copy paste. L#1482 L#1507 L#1522, extra common function

**Comment by lijinke666** [comment]
Thank you. some need todo:
- [ ] add example

eg.
https://github.com/lijinke666/react-music-player/blob/7f4e19ef60752f4f37718ea361068c33892e90cd/example/example.js#L763

- [ ] add test case
- [ ] optimize code
- [ ] README api doc

**Comment by aniketbiswas21** [comment]
@lijinke666 I have made all the requested changes. Still, not so sure about the failing CodeFactor test.

File: index.d.ts [hunk 4]
```diff
@@ -234,6 +234,7 @@ export interface ReactJkMusicPlayerProps {
     fadeIn?: number
     fadeOut?: number
   }
+  restartCurrentOnPrev?: Boolean
```
Comment (lijinke666) [inline index.d.ts:237]: Boolean => boolean

File: src/index.js [hunk 5]
```diff
@@ -1490,6 +1493,17 @@ export default class ReactJkMusicPlayer extends PureComponent {
           }
           this.audioListsPlay(audioLists[currentPlayIndex + 1][PLAYER_KEY])
         } else {
+          if (restartCurrentOnPrev) {
```
Comment (lijinke666) [inline src/index.js:1496]: restartCurrentOnPrev only for orderLoop play mode ?

File: src/index.js [hunk 6]
```diff
@@ -1490,6 +1493,17 @@ export default class ReactJkMusicPlayer extends PureComponent {
...
+            if (this.audio.currentTime <= 1) {
```
Comment (lijinke666) [inline src/index.js:1497]: Please add restartCurrentOnPrev unit test case in `__tests__/tests/player.test.js`

Comment (aniketbiswas21) [reply src/index.js:1497]: Actually, I don't have much experience writing tests. Can you help me out with this?

**Comment by aniketbiswas21** [comment]
@lijinke666 I have made the changes but can you help me out with the tests?

**Comment by aniketbiswas21** [comment]
@lijinke666 Ma'am is it fine now?

File: src/index.js [hunk 7]
```diff
@@ -1475,6 +1477,17 @@ export default class ReactJkMusicPlayer extends PureComponent {
           this.audio.pause()
           return
         }
+
+        if (restartCurrentOnPrev && !isNext) {
```
Comment (lijinke666) [inline src/index.js:1481]: `handlePlay` is common function, used by `audio reload`, `audio error play next audio` and `audio deleted`, so, do not modify this public method, you should change `onPlayPrevAudio` function, only affects the prev audio 

https://github.com/lijinke666/react-music-player/blob/7f4e19ef60752f4f37718ea361068c33892e90cd/src/index.js#L1553-L1555

like this

```tsx
  onPlayPrevAudio = () => {
    if(this.props.restartCurrentOnPrev && this.audio.currentTime > 1) {
      this.audio.currentTime = 0
     return
   }

    this.audioPrevAndNextBasePlayHandle(false)
  }
```
do you think so :) 😆

After you modify, I will immediately merge this PR. If you can't write unit test, I will write it by myself, thank you for your contribution

**Comment by aniketbiswas21** [comment]
@lijinke666 Made the requested changes. Please check :smile:

**Comment by aniketbiswas21** [comment]
@lijinke666 Thanks for helping me out :smile: 

Had a great time working on this.

Just had a question. Can you let me know when this version releases?

**Comment by lijinke666** [comment]
When I add the unit tests, Maybe released tomorrow.

**Comment by aniketbiswas21** [comment]
> When I add the unit tests, Maybe released tomorrow.

Great! Thanks! 

Love the player btw :)

**Pull request merged**